File: Sliding Window/1358.Number of Substrings Containing All Three Characters/Solution.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        cnt = [0] * 3
        n = len(s)
        ans = l = r = 0
        while r < n:
            cnt[ord(s[r]) - ord('a')] += 1
            while cnt[0] and cnt[1] and cnt[2]:
                ans += n - r
                cnt[ord(s[l]) - ord('a')] -= 1
                l += 1
            r += 1
        return ans

File: Sliding Window/1358.Number of Substrings Containing All Three Characters/Solution.java
class Solution {
    public int numberOfSubstrings(String s) {
        int[] cnt = new int[3];
        int n = s.length(), ans = 0, l = 0, r = 0;
        while (r < n) {
            cnt[s.charAt(r) - 'a']++;
            while (cnt[0] > 0 && cnt[1] > 0 && cnt[2] > 0) {
                ans += n - r;
                cnt[s.charAt(l++) - 'a']--;
            }
            r++;
        }
        return ans;
    }
}